### PR TITLE
feat(federation): Implement notifications for mentions, reply and full

### PR DESCRIPTION
**Pull request by nickvergessen**
### 🚧 Tasks

- [x] Notifications when being mentioned
- [x] Notification when being replied to
- [x] Notification when being full subscribed

### 🏁 Checklist

- [x] ⛑️ Tests (unit and/or integration) are included or not possible
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [x] 🔖 Capability is added or not needed

### CODE REVIEW

File: lib/Notification/FederationChatNotifier.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	protected function createNotification(Room $chat, ProxyCacheMessage $message, string $subject, array $subjectData = []): INotification {
+		$subjectData['userType'] = $message->getActorType();
+		$subjectData['userId'] = $message->getActorId();
+
+		$notification = $this->notificationManager->createNotification();
+		$notification
+			->setApp('spreed')
+			->setObject('chat', $chat->getToken())
+			->setSubject($subject, $subjectData)
+			->setMessage($message->getMessageType(), [
```
Comment (Antreesy) [inline lib/Notification/FederationChatNotifier.php:132]: Display name for conversation seems incorrect
image
image
image

File: lib/Notification/FederationChatNotifier.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+			} elseif ($this->isMentioned($participant, $message)) {
+				$notification = $this->createNotification($room, $message, 'mention');
+				$notification->setUser($participant->getAttendee()->getActorId());
+				$this->notificationManager->notify($notification);
+			} elseif ($this->isMentionedAll($room, $message)) {
+				$notification = $this->createNotification($room, $message, 'mention_all');
+				$notification->setUser($participant->getAttendee()->getActorId());
+				$this->notificationManager->notify($notification);
+			}
+		} elseif ($participant->getAttendee()->getNotificationLevel() === Participant::NOTIFY_ALWAYS
```
Comment (Antreesy) [inline lib/Notification/FederationChatNotifier.php:75]: POST to `api/v4/room/{token}/notify` not implemented yet, returns 406

Comment (nickvergessen) [reply lib/Notification/FederationChatNotifier.php:75]: Yeah, different todo

File: lib/Notification/FederationChatNotifier.php (hunk 1 above)
Comment (nickvergessen) [reply lib/Notification/FederationChatNotifier.php:132]: Fixed (you need to send a new message to try it)

File: lib/Model/ProxyCacheMessage.php [hunk 3]
```diff
@@ -100,11 +119,12 @@ public function jsonSerialize(): array {
 			'actorType' => $this->getActorType(),
 			'actorId' => $this->getActorId(),
 			'actorDisplayName' => $this->getActorDisplayName(),
+			'timestamp' => $this->getCreationDatetime()->getTimestamp(),
```
Comment (Antreesy) [inline lib/Model/ProxyCacheMessage.php:122]:
```xml
 <meta>
  <status>failure</status>
  <statuscode>500</statuscode>
  <message>Internal Server Error
Call to a member function getTimestamp() on null in file '/var/www/html/apps-extra/spreed/lib/Model/ProxyCacheMessage.php' line 122</message>
 </meta>
 ```
 
 happens on first rooms fetch

File: lib/Notification/FederationChatNotifier.php (hunk 1 above)
Comment (Antreesy) [reply lib/Notification/FederationChatNotifier.php:132]: Tested, perfect now

File: lib/Model/ProxyCacheMessage.php (hunk 3 above)
Comment (Antreesy) [reply lib/Model/ProxyCacheMessage.php:122]: Works after resetting

**Pull request merged**